<div align="center">

# 🌍 Global Scholarly Migration Analysis
### *Tracking the Brain Drain: A Data-Driven Journey Through Academic Migration (1996–2023)*

</div>

<div style="background-color:#1B4332; padding:20px; border-radius:10px; color:white;">

### 💭 Every year, thousands of researchers pack their bags and cross borders — chasing better funding, better labs, better opportunities.

**But where do they actually go? Who's winning this race for talent — and who's paying the price?**

</div>

<br>

<table>
<tr style="background-color:#40916C; color:white;">
<td align="center" width="25%"><h2>210</h2>Countries Analyzed</td>
<td align="center" width="25%"><h2>28</h2>Years of Data</td>
<td align="center" width="25%"><h2>142K+</h2>Migration Records</td>
<td align="center" width="25%"><h2>1</h2>Story Worth Telling</td>
</tr>
</table>

<br>

## 📌 About This Project

This project digs into **28 years of global scholarly migration data (1996–2023)**, tracking real bibliometric records from Scopus — not outdated surveys — to uncover the hidden patterns behind academic **"brain drain"** and **"brain gain."**

Every migration in this dataset is a real researcher who changed their published affiliation from one country to another. Multiply that by hundreds of thousands, across two decades, and a striking picture emerges: **a handful of countries are quietly winning the global race for scientific talent, while others are steadily losing it.**

<br>

<div style="background-color:#D8F3DC; padding:15px; border-radius:10px;">

### 🎯 The Questions Driving This Analysis

| # | Question |
|---|---|
| 🌐 | Which countries lose the most researchers — and which gain the most? |
| 🔗 | What are the world's biggest "migration corridors" between countries? |
| 💰 | Does a country's economic level affect how many scientists it keeps or loses? |
| 📈 | How have migration patterns shifted over nearly three decades? |

</div>

<br>

## ⚡ A Sneak Peek at What We Found

> The United States alone gained over **32,000 researchers** net — nearly 4x the next closest country. Meanwhile, India lost more scientific talent than any other nation on Earth. And almost every major migration corridor in the world runs directly through one country: **the United States.**

*Keep scrolling. The numbers tell a story bigger than any single headline.*

<br>

<div align="center">

🛠️ **Built with:** Python · Pandas · Streamlit · Power BI · SQLAlchemy · ERAlchemy

</div>

# **1- IMPORT LIBRARIES**

In [1]:
# ══════════════════════════════════════════════════════════
# 📦 INSTALLING REQUIRED LIBRARIES
# ══════════════════════════════════════════════════════════

# 🗺️ System-level dependency required for rendering ERD diagrams visually
!apt-get install graphviz graphviz-dev -y -q

# 🌍 For unifying country names/codes across datasets
!pip install country_converter -q

# 🗂️ For reading Parquet files (faster than Excel)
!pip install pyarrow -q

# 📐 For ORM mapping, data modeling, and ERD generation
!pip install eralchemy2 sqlalchemy pandas -q

# 🖥️ For the interactive Python dashboard
!pip install streamlit -q

# 🔍 For visualizing missing data patterns
!pip install missingno -q

print("✅ All libraries installed successfully!")

Reading package lists...
Building dependency tree...
Reading state information...
graphviz is already the newest version (2.42.2-6ubuntu0.1).
The following additional packages will be installed:
  libatk1.0-0 libatk1.0-data libgail-common libgail18 libgtk2.0-0
  libgtk2.0-bin libgtk2.0-common libgvc6-plugins-gtk librsvg2-common
  libxcomposite1 libxdot4
Suggested packages:
  gvfs
The following NEW packages will be installed:
  libatk1.0-0 libatk1.0-data libgail-common libgail18 libgraphviz-dev
  libgtk2.0-0 libgtk2.0-bin libgtk2.0-common libgvc6-plugins-gtk
  librsvg2-common libxcomposite1 libxdot4
0 upgraded, 12 newly installed, 0 to remove and 1 not upgraded.
Need to get 2,496 kB of archives.
After this operation, 7,963 kB of additional disk space will be used.
Get:1 http://archive.ubuntu.com/ubuntu jammy/main amd64 libatk1.0-data all 2.36.0-3build1 [2,824 B]
Get:2 http://archive.ubuntu.com/ubuntu jammy/main amd64 libatk1.0-0 amd64 2.36.0-3build1 [51.9 kB]
Get:3 http://archive.ubuntu

In [2]:
# ══════════════════════════════════════════════════════════
# 📚 IMPORTING LIBRARIES
# ══════════════════════════════════════════════════════════

# --- ⚙️ Built-in utilities ---
import os
import warnings
warnings.filterwarnings('ignore')

# --- 🧮 Core data handling ---
import pandas as pd
import numpy as np

# --- 📊 Visualization ---
import matplotlib.pyplot as plt
import seaborn as sns
import plotly.express as px
import plotly.graph_objects as go

# --- 🔍 Missing data exploration ---
import missingno as msno

# --- 🌍 Country name/code standardization ---
import country_converter as coc

# --- 📐 Data modeling & ERD ---
from sqlalchemy import create_engine, Column, Integer, String, Float, ForeignKey
from sqlalchemy.orm import declarative_base, relationship
from eralchemy2 import render_er

# --- 📈 Statistics (for GDP vs migration correlation) ---
from scipy import stats

# --- 🎨 Plot styling ---
sns.set_style("whitegrid")
plt.rcParams['figure.figsize'] = (12, 6)

print("✅ All libraries imported and ready to go!")

✅ All libraries imported and ready to go!


# **2- LOADING DATASETS**

In [3]:
# Loading the datasets from a specific path
df1 = pd.read_csv("/content/dfcountries.csv")

In [4]:
df2 = pd.read_csv("/content/scopus_2024_V1_scholarlymigration_country.csv")

In [5]:
df3 = pd.read_csv("/content/scopus_2024_V1_scholarlymigration_countryflows.csv")

In [6]:
df4 = pd.read_csv("/content/world_bank_popuplation_ffilled.csv")

# **3- BASIC INFORMATION ABOUT DATASETS**


# **1)DF1**

In [7]:
# Summary statistics for numerical columns (mean, min, max, etc.)
df1.describe()

,longitude,latitude
count,211.000000,211.000000
mean,19.139549,18.889009
std,70.391069,24.210877
min,-175.216000,-41.286500
25%,-14.440200,4.392035
50%,19.259500,17.300000
75%,50.214300,39.882400
max,179.089567,64.183600


In [8]:
# Check the number of rows and columns (dimensions of the dataframe)
df1.shape

(296, 18)

In [9]:
# Display dataframe summary, including data types and missing values
df1.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 296 entries, 0 to 295
Data columns (total 18 columns):
 #   Column                Non-Null Count  Dtype  
---  ------                --------------  -----  
 0   iso3code              296 non-null    object 
 1   iso2Code              295 non-null    object 
 2   name                  296 non-null    object 
 3   capitalCity           211 non-null    object 
 4   longitude             211 non-null    float64
 5   latitude              211 non-null    float64
 6   region.id             217 non-null    object 
 7   region.iso2code       217 non-null    object 
 8   region_value          296 non-null    object 
 9   adminregion.id        134 non-null    object 
 10  adminregion.iso2code  134 non-null    object 
 11  adminregion.value     134 non-null    object 
 12  incomeLevel_id        217 non-null    object 
 13  incomeLevel.iso2code  217 non-null    object 
 14  incomeLevel.value     296 non-null    object 
 15  lendingType.id        2

In [10]:
# Preview the first 5 rows of the dataset
df1.head()

,iso3code,iso2Code,name,capitalCity,longitude,latitude,region.id,region.iso2code,region_value,adminregion.id,adminregion.iso2code,adminregion.value,incomeLevel_id,incomeLevel.iso2code,incomeLevel.value,lendingType.id,lendingType.iso2code,lendingType.value
0,ABW,AW,Aruba,Oranjestad,-70.0167,12.5167,LCN,ZJ,Latin America & Caribbean,NaN,NaN,NaN,HIC,XD,High income,LNX,XX,Not classified
1,AFE,ZH,Africa Eastern and Southern,NaN,NaN,NaN,NaN,NaN,Aggregates,NaN,NaN,NaN,NaN,NaN,Aggregates,NaN,NaN,Aggregates
2,AFG,AF,Afghanistan,Kabul,69.1761,34.5228,SAS,8S,South Asia,SAS,8S,South Asia,LIC,XM,Low income,IDX,XI,IDA
3,AFR,A9,Africa,NaN,NaN,NaN,NaN,NaN,Aggregates,NaN,NaN,NaN,NaN,NaN,Aggregates,NaN,NaN,Aggregates
4,AFW,ZI,Africa Western and Central,NaN,NaN,NaN,NaN,NaN,Aggregates,NaN,NaN,NaN,NaN,NaN,Aggregates,NaN,NaN,Aggregates


In [11]:
# Preview the last 5 rows of the dataset
df1.tail()

,iso3code,iso2Code,name,capitalCity,longitude,latitude,region.id,region.iso2code,region_value,adminregion.id,adminregion.iso2code,adminregion.value,incomeLevel_id,incomeLevel.iso2code,incomeLevel.value,lendingType.id,lendingType.iso2code,lendingType.value
291,XZN,A5,Sub-Saharan Africa excluding South Africa and ...,NaN,NaN,NaN,NaN,NaN,Aggregates,NaN,NaN,NaN,NaN,NaN,Aggregates,NaN,NaN,Aggregates
292,YEM,YE,"Yemen, Rep.",Sana'a,44.2075,15.3520,MEA,ZQ,Middle East & North Africa,MNA,XQ,Middle East & North Africa (excluding high inc...,LIC,XM,Low income,IDX,XI,IDA
293,ZAF,ZA,South Africa,Pretoria,28.1871,-25.7460,SSF,ZG,Sub-Saharan Africa,SSA,ZF,Sub-Saharan Africa (excluding high income),UMC,XT,Upper middle income,IBD,XF,IBRD
294,ZMB,ZM,Zambia,Lusaka,28.2937,-15.3982,SSF,ZG,Sub-Saharan Africa,SSA,ZF,Sub-Saharan Africa (excluding high income),LMC,XN,Lower middle income,IDX,XI,IDA
295,ZWE,ZW,Zimbabwe,Harare,31.0672,-17.8312,SSF,ZG,Sub-Saharan Africa,SSA,ZF,Sub-Saharan Africa (excluding high income),LMC,XN,Lower middle income,IDB,XH,Blend


# **2)DF2**

In [12]:
df2.describe()

,year,padded_population_of_researchers,paddedpop,number_of_inmigrations,number_of_outmigrations,netmigration,outmigrationrate,inmigrationrate,netmigrationrate
count,5675.000000,5.675000e+03,5.675000e+03,5675.000000,5675.000000,5675.000000,5365.000000,5365.000000,5365.000000
mean,2009.519295,2.496270e+04,2.496270e+04,442.497269,442.503084,-0.005815,0.049236,0.054100,0.004864
std,7.918498,1.155877e+05,1.155877e+05,1637.018900,1596.243089,223.523840,0.073307,0.058373,0.074750
min,1996.000000,0.000000e+00,0.000000e+00,0.000000,0.000000,-2198.000000,0.000000,0.000000,-2.000000
25%,2003.000000,6.750000e+01,6.750000e+01,6.000000,5.000000,-3.000000,0.016159,0.017699,-0.003373
50%,2010.000000,5.780000e+02,5.780000e+02,32.000000,28.000000,1.000000,0.030612,0.036913,0.001342
75%,2016.000000,6.809000e+03,6.809000e+03,194.000000,176.000000,14.000000,0.054159,0.069444,0.016949
max,2023.000000,2.251676e+06,2.251676e+06,24990.000000,23310.000000,3276.000000,2.000000,0.666667,0.571429


In [13]:
df2.shape

(5675, 10)

In [14]:
df2.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 5675 entries, 0 to 5674
Data columns (total 10 columns):
 #   Column                            Non-Null Count  Dtype  
---  ------                            --------------  -----  
 0   year                              5675 non-null   float64
 1   countrycode                       5675 non-null   object 
 2   padded_population_of_researchers  5675 non-null   int64  
 3   paddedpop                         5675 non-null   int64  
 4   number_of_inmigrations            5675 non-null   int64  
 5   number_of_outmigrations           5675 non-null   int64  
 6   netmigration                      5675 non-null   int64  
 7   outmigrationrate                  5365 non-null   float64
 8   inmigrationrate                   5365 non-null   float64
 9   netmigrationrate                  5365 non-null   float64
dtypes: float64(4), int64(5), object(1)
memory usage: 443.5+ KB


In [15]:
df2.head()

,year,countrycode,padded_population_of_researchers,paddedpop,number_of_inmigrations,number_of_outmigrations,netmigration,outmigrationrate,inmigrationrate,netmigrationrate
0,1997.0,ABW,1,1,0,0,0,0.0,0.000000,0.000000
1,2000.0,ABW,2,2,0,0,0,0.0,0.000000,0.000000
2,2009.0,ABW,9,9,0,0,0,0.0,0.000000,0.000000
3,2010.0,ABW,11,11,1,0,1,0.0,0.090909,0.090909
4,2011.0,ABW,14,14,0,0,0,0.0,0.000000,0.000000


In [16]:
df2.tail()

,year,countrycode,padded_population_of_researchers,paddedpop,number_of_inmigrations,number_of_outmigrations,netmigration,outmigrationrate,inmigrationrate,netmigrationrate
5670,2019.0,ZWE,2048,2048,106,93,13,0.045410,0.051758,0.006348
5671,2020.0,ZWE,2083,2083,89,99,-10,0.047528,0.042727,-0.004801
5672,2021.0,ZWE,2024,2024,113,121,-8,0.059783,0.055830,-0.003953
5673,2022.0,ZWE,0,0,82,99,-17,NaN,NaN,NaN
5674,2023.0,ZWE,0,0,47,69,-22,NaN,NaN,NaN


# **3)DF3**

In [17]:
df3.describe()

,year,n_migrations,n_migrations_back,netmigrations
count,142825.000000,1.428250e+05,142825.000000,1.428250e+05
mean,2011.906382,2.161558e+02,17.193299,1.989625e+02
std,7.592905,5.052673e+03,89.992466,5.052598e+03
min,1994.000000,1.000000e+00,0.000000,-1.937000e+03
25%,2006.000000,1.000000e+00,0.000000,-1.000000e+00
50%,2013.000000,3.000000e+00,2.000000,1.000000e+00
75%,2018.000000,1.000000e+01,7.000000,2.000000e+00
max,2026.000000,1.163994e+06,4508.000000,1.163994e+06


In [18]:
df3.shape

(142825, 6)

In [19]:
df3.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 142825 entries, 0 to 142824
Data columns (total 6 columns):
 #   Column             Non-Null Count   Dtype  
---  ------             --------------   -----  
 0   migrationfrom      137150 non-null  object 
 1   migrationto        137408 non-null  object 
 2   year               142825 non-null  float64
 3   n_migrations       142825 non-null  int64  
 4   n_migrations_back  142825 non-null  int64  
 5   netmigrations      142825 non-null  int64  
dtypes: float64(1), int64(3), object(2)
memory usage: 6.5+ MB


In [20]:
df3.head()

,migrationfrom,migrationto,year,n_migrations,n_migrations_back,netmigrations
0,USA,DEU,2014.0,1270,1372,-102
1,USA,SAU,2012.0,131,58,73
2,USA,CHN,2006.0,859,1109,-250
3,USA,CAN,2020.0,1851,2045,-194
4,JPN,USA,2003.0,1634,1356,278


In [21]:
df3.tail()

,migrationfrom,migrationto,year,n_migrations,n_migrations_back,netmigrations
142820,SVK,POL,2005.0,1,1,0
142821,CZE,IRN,2006.0,1,1,0
142822,RUS,TWN,2023.0,1,2,-1
142823,HRV,RUS,2018.0,1,1,0
142824,OMN,JPN,2018.0,1,2,-1


# **4)DF4**

In [22]:
df4.describe()

,date,value
count,7336.000000,7.336000e+03
mean,2009.500000,2.635426e+08
std,8.078298,8.672790e+08
min,1996.000000,9.609000e+03
25%,2002.750000,1.361900e+06
50%,2009.500000,9.075626e+06
75%,2016.250000,5.109702e+07
max,2023.000000,7.950947e+09


In [23]:
df4.shape

(7336, 3)

In [24]:
df4.info()

<class 'pandas.core.frame.DataFrame'>
RangeIndex: 7336 entries, 0 to 7335
Data columns (total 3 columns):
 #   Column           Non-Null Count  Dtype  
---  ------           --------------  -----  
 0   countryiso3code  7308 non-null   object 
 1   date             7336 non-null   int64  
 2   value            7336 non-null   float64
dtypes: float64(1), int64(1), object(1)
memory usage: 172.1+ KB


In [25]:
df4.head()

,countryiso3code,date,value
0,NaN,1996,1.447252e+09
1,ABW,1996,7.941700e+04
2,AFE,1996,3.629858e+08
3,AFG,1996,1.710660e+07
4,AFW,1996,2.422003e+08


In [26]:
df4.tail()

,countryiso3code,date,value
7331,XKX,2023,1761985.0
7332,YEM,2023,33696614.0
7333,ZAF,2023,59893885.0
7334,ZMB,2023,20017675.0
7335,ZWE,2023,16320537.0


# **3- Data Quality Assessment**

In [ ]:
import pandas as pd
import numpy as np
from difflib import SequenceMatcher
from itertools import combinations


# --------------------------------------------------------------------
# Code 1: Missing Values Check
# Calculates the count and percentage of NaN values in each column,
# and sorts them from highest to lowest
# --------------------------------------------------------------------
def missing_values_report(df: pd.DataFrame, name: str = "") -> pd.DataFrame:
    print(f"\n{'='*70}\n1) MISSING VALUES REPORT | {name}\n{'='*70}")
    miss_count = df.isna().sum()
    miss_pct = (miss_count / len(df) * 100).round(2)
    report = pd.DataFrame({
        "missing_count": miss_count,
        "missing_pct": miss_pct
    }).sort_values("missing_count", ascending=False)
    result = report[report.missing_count > 0]
    if result.empty:
        print("No missing values ✅")
    else:
        print(result)
    return result


# --------------------------------------------------------------------
# Code 2: Duplicates Check
# Checks for fully duplicated rows, and also partial duplicates based
# on a logical key (e.g., the same country and same year appearing
# twice, which shouldn't happen)
# --------------------------------------------------------------------
def duplicates_report(df: pd.DataFrame, name: str = "", key_cols: list = None) -> None:
    print(f"\n{'='*70}\n2) DUPLICATES REPORT | {name}\n{'='*70}")
    full_dupes = df.duplicated().sum()
    print(f"Fully duplicated rows (all columns): {full_dupes}")

    if key_cols:
        key_dupes = df.duplicated(subset=key_cols).sum()
        print(f"Rows duplicated on key {key_cols}: {key_dupes}")
        if key_dupes > 0:
            print("Sample of rows duplicated on the key:")
            print(df[df.duplicated(subset=key_cols, keep=False)].sort_values(key_cols).head(10))


# --------------------------------------------------------------------
# Code 3: Data Types Check
# Prints the type of each column, and tries to detect if a numeric
# column was stored as text (object), or if a date column is still
# a string, so it can be converted before any analysis
# --------------------------------------------------------------------
def dtype_report(df: pd.DataFrame, name: str = "") -> None:
    print(f"\n{'='*70}\n3) DATA TYPES REPORT | {name}\n{'='*70}")
    print(df.dtypes)

    # Detect object columns that look like numbers stored incorrectly as text
    suspects = []
    for col in df.select_dtypes(include=["object"]).columns:
        sample = df[col].dropna().astype(str).head(50)
        numeric_like = sample.str.replace(".", "", regex=False).str.replace("-", "", regex=False).str.isnumeric()
        if numeric_like.mean() > 0.8 and len(sample) > 0:
            suspects.append(col)
    if suspects:
        print(f"\n⚠️ Text (object) columns that look like numbers: {suspects}")
    else:
        print("\nNo text columns suspected of being numbers ✅")


# --------------------------------------------------------------------
# Code 4: Outliers Check using the IQR method
# Calculates the normal bounds for each numeric column (Q1 - 1.5*IQR)
# and (Q3 + 1.5*IQR), and reports the count and percentage of values
# that fall outside these bounds
# --------------------------------------------------------------------
def outliers_report(df: pd.DataFrame, name: str = "", cols: list = None) -> pd.DataFrame:
    print(f"\n{'='*70}\n4) OUTLIERS REPORT (IQR method) | {name}\n{'='*70}")
    numeric_cols = cols if cols else df.select_dtypes(include=np.number).columns.tolist()
    rows = []
    for col in numeric_cols:
        series = df[col].dropna()
        if series.empty:
            continue
        q1, q3 = series.quantile(0.25), series.quantile(0.75)
        iqr = q3 - q1
        lower, upper = q1 - 1.5 * iqr, q3 + 1.5 * iqr
        outliers = series[(series < lower) | (series > upper)]
        rows.append({
            "column": col,
            "lower_bound": round(lower, 2),
            "upper_bound": round(upper, 2),
            "outliers_count": len(outliers),
            "outliers_pct": round(len(outliers) / len(series) * 100, 2)
        })
    result = pd.DataFrame(rows).sort_values("outliers_count", ascending=False)
    print(result)
    return result


# --------------------------------------------------------------------
# Code 5: Whitespace Check (leading/trailing spaces)
# Checks every text column for values with a space at the start or
# end (e.g., " Egypt" or "Egypt " instead of "Egypt"), since these
# break any Group By or Join
# --------------------------------------------------------------------
def whitespace_report(df: pd.DataFrame, name: str = "") -> None:
    print(f"\n{'='*70}\n5) WHITESPACE CHECK (leading/trailing spaces) | {name}\n{'='*70}")
    found_any = False
    for col in df.select_dtypes(include=["object"]).columns:
        mask = df[col].dropna().astype(str).apply(lambda x: x != x.strip())
        count = mask.sum()
        if count > 0:
            found_any = True
            print(f"Column '{col}': {count} value(s) with extra leading/trailing whitespace")
            examples = df[col].dropna().astype(str)[mask].unique()[:5]
            print(f"   Examples: {[repr(e) for e in examples]}")
    if not found_any:
        print("No leading/trailing whitespace found in any text ✅")


# --------------------------------------------------------------------
# Code 6: Inconsistent Text / Fuzzy Duplicates Check
# Looks for values in the same column that are very similar but
# written differently (e.g., "Egypt" vs "egypt" vs "EGYPT ", or
# "USA" vs "U.S.A") using:
#   (a) Normalizing case and whitespace as a quick first pass
#   (b) Text similarity ratio (SequenceMatcher) to detect near matches
# --------------------------------------------------------------------
def inconsistent_text_report(df: pd.DataFrame, name: str = "", cols: list = None,
                              similarity_threshold: float = 0.85) -> None:
    print(f"\n{'='*70}\n6) INCONSISTENT TEXT REPORT | {name}\n{'='*70}")
    text_cols = cols if cols else df.select_dtypes(include=["object"]).columns.tolist()

    for col in text_cols:
        values = df[col].dropna().astype(str).unique()
        if len(values) < 2 or len(values) > 2000:  # skip columns with too many unique values (e.g., unique names)
            continue

        # Step (a): was the same value written with different case/whitespace?
        normalized_map = {}
        for v in values:
            norm = v.strip().lower()
            normalized_map.setdefault(norm, []).append(v)
        case_conflicts = {k: v for k, v in normalized_map.items() if len(v) > 1}
        if case_conflicts:
            print(f"\nColumn '{col}' -> same value in different forms (case/whitespace):")
            for norm, variants in case_conflicts.items():
                print(f"   {variants}")

        # Step (b): high text similarity between values that are written completely differently
        # (limiting the number of comparisons for performance when there are many values)
        if len(values) <= 300:
            similar_pairs = []
            for v1, v2 in combinations(values, 2):
                if v1.strip().lower() == v2.strip().lower():
                    continue  # already caught above
                ratio = SequenceMatcher(None, v1.lower(), v2.lower()).ratio()
                if ratio >= similarity_threshold:
                    similar_pairs.append((v1, v2, round(ratio, 2)))
            if similar_pairs:
                print(f"\nColumn '{col}' -> very similar values that might mean the same thing:")
                for v1, v2, ratio in similar_pairs:
                    print(f"   '{v1}'  <->  '{v2}'   (similarity: {ratio})")


# --------------------------------------------------------------------
# Loading the files (change the paths here based on where your files are)
# --------------------------------------------------------------------
PATH_POPULATION      = "/content/world_bank_popuplation_ffilled.csv"
PATH_COUNTRIES       = "/content/dfcountries.csv"
PATH_SCOPUS_COUNTRY  = "/content/scopus_2024_V1_scholarlymigration_country.csv"
PATH_SCOPUS_FLOWS    = "/content/scopus_2024_V1_scholarlymigration_countryflows.csv"


# ==========================================================================
#  TABLE 1: POPULATION  (world_bank_popuplation_ffilled)
#  Grain: country + year
# ==========================================================================
def run_checks_population():
    df = pd.read_csv(PATH_POPULATION)
    name = "population"
    print(f"\n\n{'#'*70}\n#  TABLE 1: {name}\n{'#'*70}")

    missing_values_report(df, name)
    duplicates_report(df, name, key_cols=["countryiso3code", "date"])
    dtype_report(df, name)
    outliers_report(df, name)
    whitespace_report(df, name)
    inconsistent_text_report(df, name)
    return df


# ==========================================================================
#  TABLE 2: COUNTRIES  (dfcountries) -> Dimension table
#  Grain: country
# ==========================================================================
def run_checks_countries():
    df = pd.read_csv(PATH_COUNTRIES)
    name = "countries"
    print(f"\n\n{'#'*70}\n#  TABLE 2: {name}\n{'#'*70}")

    missing_values_report(df, name)
    duplicates_report(df, name, key_cols=["iso3code"])
    dtype_report(df, name)
    outliers_report(df, name)
    whitespace_report(df, name)
    inconsistent_text_report(df, name)
    return df


# ==========================================================================
#  TABLE 3: SCOPUS_COUNTRY  (scopus_2024_V1_scholarlymigration_country)
#  Grain: country + year
# ==========================================================================
def run_checks_scopus_country():
    df = pd.read_csv(PATH_SCOPUS_COUNTRY)
    name = "scopus_country"
    print(f"\n\n{'#'*70}\n#  TABLE 3: {name}\n{'#'*70}")

    missing_values_report(df, name)
    duplicates_report(df, name, key_cols=["countrycode", "year"])
    dtype_report(df, name)
    outliers_report(df, name)
    whitespace_report(df, name)
    inconsistent_text_report(df, name)
    return df


# ==========================================================================
#  TABLE 4: SCOPUS_FLOWS  (scopus_2024_V1_scholarlymigration_countryflows)
#  Grain: from_country + to_country + year
# ==========================================================================
def run_checks_scopus_flows():
    df = pd.read_csv(PATH_SCOPUS_FLOWS)
    name = "scopus_flows"
    print(f"\n\n{'#'*70}\n#  TABLE 4: {name}\n{'#'*70}")

    missing_values_report(df, name)
    duplicates_report(df, name, key_cols=["migrationfrom", "migrationto", "year"])
    dtype_report(df, name)
    outliers_report(df, name)
    whitespace_report(df, name)
    inconsistent_text_report(df, name)
    return df


# --------------------------------------------------------------------
# Execution: each table runs independently, fully separate from the rest
# If you want to run just one table, keep its line and comment out the rest
# --------------------------------------------------------------------
if __name__ == "__main__":
    df_population     = run_checks_population()
    df_countries      = run_checks_countries()
    df_scopus_country = run_checks_scopus_country()
    df_scopus_flows   = run_checks_scopus_flows()


# **4- Data Cleaning**

In [ ]:
import pandas as pd
import numpy as np
import os

# --------------------------------------------------------------------
# File paths (change these based on where your files are)
# --------------------------------------------------------------------
PATH_POPULATION      = "/content/world_bank_popuplation_ffilled.csv"
PATH_COUNTRIES       = "/content/dfcountries.csv"
PATH_SCOPUS_COUNTRY  = "/content/scopus_2024_V1_scholarlymigration_country.csv"
PATH_SCOPUS_FLOWS    = "/content/scopus_2024_V1_scholarlymigration_countryflows.csv"

OUTPUT_DIR = "/mnt/user-data/outputs"

# --------------------------------------------------------------------
# Aggregate codes found in the fact tables (not in the original countries
# table). These are statistical aggregations (World, Regions, Income
# groups) that also need to be removed from any fact table so they
# don't break the join with Dim_Country
# --------------------------------------------------------------------
AGGREGATE_CODES_TO_EXCLUDE = [
    'AFE', 'AFW', 'ARB', 'CEB', 'CSS', 'EAP', 'EAR', 'EAS', 'ECA', 'ECS',
    'EMU', 'EUU', 'FCS', 'HPC', 'IBD', 'IBT', 'IDA', 'IDB', 'IDX', 'LAC',
    'LCN', 'LDC', 'LMY', 'LTE', 'MEA', 'MIC', 'MNA', 'NAC', 'OED', 'OSS',
    'PRE', 'PSS', 'PST', 'SAS', 'SSA', 'SSF', 'SST', 'TEA', 'TEC', 'TLA',
    'TMN', 'TSA', 'TSS', 'WLD',
]

# --------------------------------------------------------------------
# Real entities (countries/territories) that exist in the Scopus data
# but are not originally present in the dfcountries.csv file (because
# the World Bank doesn't cover them) - we add them manually with basic
# info so the join is complete and there are no orphan rows
# --------------------------------------------------------------------
MANUAL_MISSING_ENTITIES = [
    # iso3, name,                              region_name
    ("AIA", "Anguilla",                          "Latin America & Caribbean"),
    ("ATA", "Antarctica",                        "Not classified"),
    ("ATF", "French Southern Territories",       "Not classified"),
    ("BES", "Bonaire, Sint Eustatius and Saba",  "Latin America & Caribbean"),
    ("BVT", "Bouvet Island",                     "Not classified"),
    ("CCK", "Cocos (Keeling) Islands",           "East Asia & Pacific"),
    ("CXR", "Christmas Island",                  "East Asia & Pacific"),
    ("EUE", "Europe (unclassified)",             "Not classified"),
    ("FLK", "Falkland Islands",                  "Latin America & Caribbean"),
    ("GLP", "Guadeloupe",                        "Latin America & Caribbean"),
    ("GUF", "French Guiana",                     "Latin America & Caribbean"),
    ("IOT", "British Indian Ocean Territory",    "South Asia"),
    ("JEY", "Jersey",                            "Europe & Central Asia"),
    ("MSR", "Montserrat",                        "Latin America & Caribbean"),
    ("MTQ", "Martinique",                        "Latin America & Caribbean"),
    ("NFK", "Norfolk Island",                    "East Asia & Pacific"),
    ("NIU", "Niue",                              "East Asia & Pacific"),
    ("REU", "Réunion",                           "Sub-Saharan Africa"),
    ("SGS", "South Georgia and South Sandwich Islands", "Not classified"),
    ("SHN", "Saint Helena",                      "Sub-Saharan Africa"),
    ("SJM", "Svalbard and Jan Mayen",            "Europe & Central Asia"),
    ("SPM", "Saint Pierre and Miquelon",         "North America"),
    ("TKL", "Tokelau",                           "East Asia & Pacific"),
    ("TWN", "Taiwan",                            "East Asia & Pacific"),
    ("VAT", "Holy See (Vatican)",                "Europe & Central Asia"),
    ("WLF", "Wallis and Futuna",                 "East Asia & Pacific"),
]


# ==========================================================================
#  TABLE 1: COUNTRIES  ->  Dim_Country
#  Decisions:
#   - Remove aggregate rows (World, Africa, Income groups...) since they
#     aren't real countries and would break any join with the fact tables
#     (identified by an empty region.id)
#   - Strip extra whitespace from all text columns
#   - Drop the adminregion column since it's 54.7% empty and doesn't add
#     meaningful analytical value
#   - Sort/rename the columns cleanly and clearly as a final Dimension
# ==========================================================================
def clean_countries():
    df = pd.read_csv(PATH_COUNTRIES)
    print(f"\n{'#'*70}\n#  CLEANING TABLE 1: countries -> Dim_Country\n{'#'*70}")
    print(f"Shape before cleaning: {df.shape}")

    # 1) Strip extra whitespace from all text columns
    text_cols = df.select_dtypes(include=["object"]).columns
    for col in text_cols:
        df[col] = df[col].str.strip()

    # 2) Remove aggregate rows (empty region.id = not a real country)
    before = len(df)
    df = df[df["region.id"].notna()].copy()
    print(f"Removed {before - len(df)} aggregate row(s) (World, Regions, Income groups...)")

    # 3) Drop the adminregion column (mostly empty and not important for analysis)
    df = df.drop(columns=["adminregion.id", "adminregion.iso2code", "adminregion.value"])

    # 4) Select, order, and rename the final Dimension columns
    df = df.rename(columns={
        "iso3code": "country_iso3",
        "iso2Code": "country_iso2",
        "name": "country_name",
        "capitalCity": "capital_city",
        "region.id": "region_id",
        "region_value": "region_name",
        "incomeLevel_id": "income_level_id",
        "incomeLevel.value": "income_level_name",
        "lendingType.id": "lending_type_id",
        "lendingType.value": "lending_type_name",
    })
    df = df.drop(columns=["region.iso2code", "incomeLevel.iso2code", "lendingType.iso2code"])

    dim_country = df[[
        "country_iso3", "country_iso2", "country_name", "capital_city",
        "latitude", "longitude", "region_id", "region_name",
        "income_level_id", "income_level_name",
        "lending_type_id", "lending_type_name"
    ]].reset_index(drop=True)

    # 5) Add the missing real entities (present in Scopus, not in World Bank)
    manual_rows = pd.DataFrame(
        MANUAL_MISSING_ENTITIES, columns=["country_iso3", "country_name", "region_name"]
    )
    dim_country = pd.concat([dim_country, manual_rows], ignore_index=True)
    print(f"Added {len(manual_rows)} missing real entity/entities (Taiwan, Vatican, Guadeloupe...)")

    print(f"Shape after cleaning: {dim_country.shape}")
    print("✅ Dim_Country is ready")
    return dim_country


# ==========================================================================
#  TABLE 2: POPULATION  ->  part of Fact_Country_Year
#  Decisions:
#   - Remove rows where countryiso3code is empty (this is the World
#     aggregate, not a real country)
#   - Rename columns consistently so they merge easily with scopus_country
# ==========================================================================
def clean_population():
    df = pd.read_csv(PATH_POPULATION)
    print(f"\n{'#'*70}\n#  CLEANING TABLE 2: population\n{'#'*70}")
    print(f"Shape before cleaning: {df.shape}")

    before = len(df)
    df = df[df["countryiso3code"].notna()].copy()
    print(f"Removed {before - len(df)} row(s) (empty countryiso3code = World aggregate)")

    df = df.rename(columns={
        "countryiso3code": "country_iso3",
        "date": "year",
        "value": "population"
    })

    print(f"Shape after cleaning: {df.shape}")
    print("✅ population is ready to merge")
    return df


# ==========================================================================
#  TABLE 3: SCOPUS_COUNTRY  ->  part of Fact_Country_Year
#  Decisions:
#   - Leave the nulls in the rate columns as they are (known reason: the
#     2022/2023 years haven't been updated yet) - don't fill with zero,
#     since zero would mean "migration rate = zero," which is wrong, not
#     "unknown"
#   - Convert year from float to int so the merge with population is clean
# ==========================================================================
def clean_scopus_country():
    df = pd.read_csv(PATH_SCOPUS_COUNTRY)
    print(f"\n{'#'*70}\n#  CLEANING TABLE 3: scopus_country\n{'#'*70}")
    print(f"Shape before cleaning: {df.shape}")

    df["year"] = df["year"].astype(int)

    df = df.rename(columns={
        "countrycode": "country_iso3",
    })

    print(f"Shape after cleaning: {df.shape}")
    print(f"Null values in the rate columns were left intentionally ({df['netmigrationrate'].isna().sum()} rows) - reason: data not yet updated")
    print("✅ scopus_country is ready to merge")
    return df


# ==========================================================================
#  TABLE 4: SCOPUS_FLOWS  ->  Fact_Migration_Flows
#  Decisions:
#   - Empty values in migrationfrom/migrationto mean "the other party is
#     unknown/unclassified," not an error - we set them to 'UNKNOWN'
#     instead of NaN so they remain groupable without being dropped from
#     the join
#   - Convert year to int
# ==========================================================================
def clean_scopus_flows():
    df = pd.read_csv(PATH_SCOPUS_FLOWS)
    print(f"\n{'#'*70}\n#  CLEANING TABLE 4: scopus_flows -> Fact_Migration_Flows\n{'#'*70}")
    print(f"Shape before cleaning: {df.shape}")

    df["migrationfrom"] = df["migrationfrom"].fillna("UNKNOWN")
    df["migrationto"] = df["migrationto"].fillna("UNKNOWN")
    df["year"] = df["year"].astype(int)

    df = df.rename(columns={
        "migrationfrom": "from_country_iso3",
        "migrationto": "to_country_iso3",
    })

    # Remove rows with an aggregate code instead of a real country
    before = len(df)
    df = df[
        ~df["from_country_iso3"].isin(AGGREGATE_CODES_TO_EXCLUDE) &
        ~df["to_country_iso3"].isin(AGGREGATE_CODES_TO_EXCLUDE)
    ].copy()
    print(f"Removed {before - len(df)} row(s) with an aggregate code instead of a real country")

    print(f"Shape after cleaning: {df.shape}")
    print("✅ Fact_Migration_Flows is ready")
    return df


# ==========================================================================
#  Merge population + scopus_country into a single fact table
#  Reason: they're at exactly the same grain (country + year), so there's
#  no need to keep them as two separate tables - this reduces the number
#  of tables and simplifies the Star Schema
# ==========================================================================
def build_fact_country_year(df_population, df_scopus_country):
    print(f"\n{'#'*70}\n#  BUILDING: Fact_Country_Year (population + scopus_country)\n{'#'*70}")

    fact = pd.merge(
        df_population,
        df_scopus_country,
        on=["country_iso3", "year"],
        how="outer"   # outer so we keep any country/year present in either source
    )

    # Remove rows with an aggregate code (WLD, SSA, EAP...) instead of a real country
    before = len(fact)
    fact = fact[~fact["country_iso3"].isin(AGGREGATE_CODES_TO_EXCLUDE)].copy()
    print(f"Removed {before - len(fact)} row(s) with an aggregate code instead of a real country")

    print(f"Final fact table shape: {fact.shape}")
    print(f"Number of unique countries: {fact['country_iso3'].nunique()}")
    print(f"Year range: {fact['year'].min()} - {fact['year'].max()}")
    print("✅ Fact_Country_Year is ready")
    return fact


# ==========================================================================
#  Final check: make sure every country code in the fact tables exists
#  in Dim_Country (Referential Integrity Check) - this must be zero after
#  all the changes above
# ==========================================================================
def check_referential_integrity(dim_country, fact_country_year, fact_flows):
    print(f"\n{'#'*70}\n#  FINAL CHECK: Referential Integrity\n{'#'*70}")
    dim_codes = set(dim_country["country_iso3"].unique())

    orphan_fact = set(fact_country_year["country_iso3"].unique()) - dim_codes
    print(f"Codes in Fact_Country_Year without a Dimension: {len(orphan_fact)} {sorted(orphan_fact) if orphan_fact else ''}")

    flow_codes = set(fact_flows["from_country_iso3"].unique()) | set(fact_flows["to_country_iso3"].unique())
    orphan_flows = flow_codes - dim_codes - {"UNKNOWN"}
    print(f"Codes in Fact_Migration_Flows without a Dimension: {len(orphan_flows)} {sorted(orphan_flows) if orphan_flows else ''}")

    if not orphan_fact and not orphan_flows:
        print("✅ Referential integrity is 100% intact - every code in the fact tables exists in Dim_Country")


# ==========================================================================
#  Execution: run each table independently, then build the merged fact
#  table and save everything
# ==========================================================================
if __name__ == "__main__":
    dim_country       = clean_countries()
    df_population     = clean_population()
    df_scopus_country = clean_scopus_country()
    fact_flows        = clean_scopus_flows()

    fact_country_year = build_fact_country_year(df_population, df_scopus_country)

    check_referential_integrity(dim_country, fact_country_year, fact_flows)

    # Create the output directory if it doesn't exist
    os.makedirs(OUTPUT_DIR, exist_ok=True)

    # -------- Final save: just 3 tables, ready for modeling --------
    dim_country.to_csv(f"{OUTPUT_DIR}/Dim_Country.csv", index=False)
    fact_country_year.to_csv(f"{OUTPUT_DIR}/Fact_Country_Year.csv", index=False)
    fact_flows.to_csv(f"{OUTPUT_DIR}/Fact_Migration_Flows.csv", index=False)

    print(f"\n{'='*70}")
    print("✅ Final files saved and ready for modeling:")
    print("   - Dim_Country.csv")
    print("   - Fact_Country_Year.csv")
    print("   - Fact_Migration_Flows.csv")
    print(f"{'='*70}")


# **5- Feature Enginering**

In [ ]:
"""
==========================================================================
 Stage: Feature Engineering
==========================================================================
Goal: add new derived columns to each table independently. These columns
      help with:
       - Quick analysis in Power BI / Streamlit without complex
         calculations every time
       - Classifying countries and flows in a way that's ready for
         filtering and display

Code blocks in this file:
  1. add_features_dim_country()        -> country-level (descriptive) features
  2. add_features_fact_country_year()  -> country+year features (growth, classification)
  3. add_features_fact_flows()         -> migration-flow-level features between two countries
==========================================================================
"""

import pandas as pd
import numpy as np

PATH_DIM_COUNTRY  = "/mnt/user-data/outputs/Dim_Country.csv"
PATH_FACT_CY      = "/mnt/user-data/outputs/Fact_Country_Year.csv"
PATH_FACT_FLOWS   = "/mnt/user-data/outputs/Fact_Migration_Flows.csv"
OUTPUT_DIR        = "/mnt/user-data/outputs"


# ==========================================================================
#  TABLE 1: Dim_Country -> add country-level descriptive features
#  New columns:
#   - hemisphere            : whether the country is in the Northern or
#                             Southern hemisphere (from latitude)
#   - is_high_income        : a quick flag (True/False) instead of reading
#                             income_level_name
#   - income_level_rank     : a numeric ranking of income (1=Low ... 4=High)
#                             so it sorts correctly in any chart (the raw
#                             text sorts alphabetically, which is wrong)
# ==========================================================================
def add_features_dim_country(df: pd.DataFrame) -> pd.DataFrame:
    print(f"\n{'#'*70}\n#  FEATURE ENGINEERING: Dim_Country\n{'#'*70}")
    df = df.copy()

    # hemisphere from latitude: positive = Northern, negative = Southern
    df["hemisphere"] = np.where(
        df["latitude"].isna(), "Unknown",
        np.where(df["latitude"] >= 0, "Northern", "Southern")
    )

    # Quick flag for high income (makes filtering easier in later analysis)
    df["is_high_income"] = df["income_level_name"] == "High income"

    # Logical numeric ranking for income level instead of the wrong alphabetical order
    income_order = {
        "Low income": 1,
        "Lower middle income": 2,
        "Upper middle income": 3,
        "High income": 4,
    }
    df["income_level_rank"] = df["income_level_name"].map(income_order)

    print("Added columns: hemisphere, is_high_income, income_level_rank")
    print(f"Final shape: {df.shape}")
    return df


# ==========================================================================
#  TABLE 2: Fact_Country_Year -> add country+year level features
#  New columns:
#   - decade                  : grouping years into decades (1990s, 2000s...)
#                               for overall trend analysis
#   - population_growth_rate  : population growth rate vs. the previous
#                               year for each country
#   - researcher_density      : ratio of number of researchers to
#                               population (per 1,000)
#   - migration_status        : a simple text classification (Brain Gain /
#                               Brain Drain / Balanced) based on
#                               netmigration, instead of the reader
#                               interpreting the number themselves
# ==========================================================================
def add_features_fact_country_year(df: pd.DataFrame) -> pd.DataFrame:
    print(f"\n{'#'*70}\n#  FEATURE ENGINEERING: Fact_Country_Year\n{'#'*70}")
    df = df.copy()
    df = df.sort_values(["country_iso3", "year"]).reset_index(drop=True)

    # Group into decades to make aggregated time-based analysis easier
    df["decade"] = (df["year"] // 10 * 10).astype("Int64").astype(str) + "s"

    # Population growth rate vs. the previous year for the same country (Year-over-Year)
    df["population_growth_rate"] = df.groupby("country_iso3")["population"].pct_change().round(4)

    # Researchers per 1,000 of population (to fairly compare countries of different sizes)
    df["researcher_density_per_1000"] = (df["paddedpop"] / df["population"] * 1000).round(3)

    # Direct text classification of migration status, instead of reading and
    # interpreting the raw numbers every time
    df["migration_status"] = np.select(
        [df["netmigration"] > 0, df["netmigration"] < 0, df["netmigration"] == 0],
        ["Brain Gain", "Brain Drain", "Balanced"],
        default="Unknown"
    )

    print("Added columns: decade, population_growth_rate, researcher_density_per_1000, migration_status")
    print(f"Final shape: {df.shape}")
    return df


# ==========================================================================
#  TABLE 3: Fact_Migration_Flows -> add features at the flow-between-two-countries level
#  New columns:
#   - decade            : grouped into decades, same as the other fact table
#   - corridor          : a unified label for the migration route "FROM-TO"
#                         (e.g., "USA-DEU") so it's easy to group or display
#                         as a single label instead of two columns
#   - flow_direction    : classifies the net migration direction (gain to
#                         `to_country` / loss / balanced)
#   - flow_size_category: classifies the size of the flow itself
#                         (Small/Medium/Large) based on quartiles, to tell
#                         apart huge corridors like US-China from the rest
# ==========================================================================
def add_features_fact_flows(df: pd.DataFrame) -> pd.DataFrame:
    print(f"\n{'#'*70}\n#  FEATURE ENGINEERING: Fact_Migration_Flows\n{'#'*70}")
    df = df.copy()

    df["decade"] = (df["year"] // 10 * 10).astype("Int64").astype(str) + "s"

    df["corridor"] = df["from_country_iso3"] + "-" + df["to_country_iso3"]

    df["flow_direction"] = np.select(
        [df["netmigrations"] > 0, df["netmigrations"] < 0, df["netmigrations"] == 0],
        ["Net flow to destination", "Net flow back to origin", "Balanced"],
        default="Unknown"
    )

    # Classify flow size based on quartiles to distinguish huge corridors
    q1, q2, q3 = df["n_migrations"].quantile([0.25, 0.5, 0.75])
    df["flow_size_category"] = pd.cut(
        df["n_migrations"],
        bins=[-np.inf, q1, q2, q3, np.inf],
        labels=["Small", "Medium", "Large", "Very Large"]
    )

    print("Added columns: decade, corridor, flow_direction, flow_size_category")
    print(f"Final shape: {df.shape}")
    return df


# ==========================================================================
#  Execution: run each table independently + save
# ==========================================================================
if __name__ == "__main__":
    dim_country       = pd.read_csv(PATH_DIM_COUNTRY)
    fact_country_year = pd.read_csv(PATH_FACT_CY)
    fact_flows        = pd.read_csv(PATH_FACT_FLOWS)

    dim_country       = add_features_dim_country(dim_country)
    fact_country_year = add_features_fact_country_year(fact_country_year)
    fact_flows        = add_features_fact_flows(fact_flows)

    dim_country.to_csv(f"{OUTPUT_DIR}/Dim_Country.csv", index=False)
    fact_country_year.to_csv(f"{OUTPUT_DIR}/Fact_Country_Year.csv", index=False)
    fact_flows.to_csv(f"{OUTPUT_DIR}/Fact_Migration_Flows.csv", index=False)

    print(f"\n{'='*70}")
    print("✅ The tables were updated with the new columns and saved")
    print(f"{'='*70}")


In [30]:
dim_country.head()

,country_iso3,country_iso2,country_name,capital_city,latitude,longitude,region_id,region_name,income_level_id,income_level_name,lending_type_id,lending_type_name,hemisphere,is_high_income,income_level_rank
0,ABW,AW,Aruba,Oranjestad,12.51670,-70.0167,LCN,Latin America & Caribbean,HIC,High income,LNX,Not classified,Northern,True,4.0
1,AFG,AF,Afghanistan,Kabul,34.52280,69.1761,SAS,South Asia,LIC,Low income,IDX,IDA,Northern,False,1.0
2,AGO,AO,Angola,Luanda,-8.81155,13.2420,SSF,Sub-Saharan Africa,LMC,Lower middle income,IBD,IBRD,Southern,False,2.0
3,ALB,AL,Albania,Tirane,41.33170,19.8172,ECS,Europe & Central Asia,UMC,Upper middle income,IBD,IBRD,Northern,False,3.0
4,AND,AD,Andorra,Andorra la Vella,42.50750,1.5218,ECS,Europe & Central Asia,HIC,High income,LNX,Not classified,Northern,True,4.0


In [31]:
fact_country_year.head()

,country_iso3,year,population,padded_population_of_researchers,paddedpop,number_of_inmigrations,number_of_outmigrations,netmigration,outmigrationrate,inmigrationrate,netmigrationrate,decade,population_growth_rate,researcher_density_per_1000,migration_status
0,ABW,1996,79417.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1990s,NaN,NaN,Unknown
1,ABW,1997,81858.0,1.0,1.0,0.0,0.0,0.0,0.0,0.0,0.0,1990s,0.0307,0.012,Balanced
2,ABW,1998,84355.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1990s,0.0305,NaN,Unknown
3,ABW,1999,86867.0,NaN,NaN,NaN,NaN,NaN,NaN,NaN,NaN,1990s,0.0298,NaN,Unknown
4,ABW,2000,89101.0,2.0,2.0,0.0,0.0,0.0,0.0,0.0,0.0,2000s,0.0257,0.022,Balanced


In [32]:
fact_flows.head()

,from_country_iso3,to_country_iso3,year,n_migrations,n_migrations_back,netmigrations,decade,corridor,flow_direction,flow_size_category
0,USA,DEU,2014,1270,1372,-102,2010s,USA-DEU,Net flow back to origin,Very Large
1,USA,SAU,2012,131,58,73,2010s,USA-SAU,Net flow to destination,Very Large
2,USA,CHN,2006,859,1109,-250,2000s,USA-CHN,Net flow back to origin,Very Large
3,USA,CAN,2020,1851,2045,-194,2020s,USA-CAN,Net flow back to origin,Very Large
4,JPN,USA,2003,1634,1356,278,2000s,JPN-USA,Net flow to destination,Very Large


In [33]:
from google.colab import files
files.download(f"{OUTPUT_DIR}/Dim_Country.csv")
files.download(f"{OUTPUT_DIR}/Fact_Country_Year.csv")
files.download(f"{OUTPUT_DIR}/Fact_Migration_Flows.csv")

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

<IPython.core.display.Javascript object>

# **8- Statistics**

In [ ]:
OUTPUT_DIR = "/mnt/user-data/outputs"

"""
==========================================================================
 Code 1: Box Plot - illustrating the population distribution and outliers
==========================================================================
Goal: visually show the distribution of the population column, and
      display the outliers as separate points outside the box - this
      visually shows that there is a large natural gap between countries
      (like India/China) and the rest of the world, not a data error.

Note: we used a log scale on the axis because the values have an
      enormous range (from 9 thousand to 1.4 billion) - without it the
      plot would be too compressed and wouldn't show anything useful.
==========================================================================
"""

import pandas as pd
import matplotlib.pyplot as plt
import seaborn as sns

fact_country_year = pd.read_csv(f"{OUTPUT_DIR}/Fact_Country_Year.csv")

plt.figure(figsize=(10, 6))
sns.boxplot(x=fact_country_year["population"], color="#4C72B0")
plt.xscale("log")  # because of the huge gap between the smallest and largest country
plt.title("Population Distribution Across All Countries & Years (Box Plot)", fontsize=14)
plt.xlabel("Population (log scale)")
plt.tight_layout()
plt.savefig("population_boxplot.png", dpi=150)
plt.show()

print("✅ Saved: population_boxplot.png")


In [ ]:
OUTPUT_DIR = "/mnt/user-data/outputs"

"""
==========================================================================
 Code 2: Scatter Plot - relationship between population and net migration,
 highlighting the outlier countries
==========================================================================
Goal: show the relationship between population and netmigration for each
      country/year as one point, highlighting/enlarging the outlier
      points (like India, China, the US) to visually show that these are
      "real extreme values," not errors - and this should visually agree
      with the Box Plot above.
==========================================================================
"""

import pandas as pd
import matplotlib.pyplot as plt

fact_country_year = pd.read_csv(f"{OUTPUT_DIR}/Fact_Country_Year.csv")
dim_country       = pd.read_csv(f"{OUTPUT_DIR}/Dim_Country.csv")

# Simple merge to bring in the country name to label the outlier points
df = fact_country_year.merge(dim_country[["country_iso3", "country_name"]], on="country_iso3", how="left")
df = df.dropna(subset=["population", "netmigration"])

# Identify the outlier countries (top 5 population values) to label them on the plot
top_outliers = df.nlargest(5, "population")

plt.figure(figsize=(11, 7))
plt.scatter(df["population"], df["netmigration"], alpha=0.4, color="#4C72B0", label="All countries")
plt.scatter(top_outliers["population"], top_outliers["netmigration"], color="red", s=80, label="Largest outliers")

# Write the country name next to each outlier point
for _, row in top_outliers.iterrows():
    plt.annotate(row["country_name"], (row["population"], row["netmigration"]),
                 textcoords="offset points", xytext=(6, 6), fontsize=9)

plt.xscale("log")  # same reason as the Box Plot: huge range in values
plt.title("Population vs Net Migration (Highlighting Real Outliers)", fontsize=14)
plt.xlabel("Population (log scale)")
plt.ylabel("Net Migration (researchers)")
plt.legend()
plt.tight_layout()
plt.savefig("population_vs_migration_scatter.png", dpi=150)
plt.show()

print("✅ Saved: population_vs_migration_scatter.png")


# **6- Data Modeling**

In [ ]:
import pandas as pd
from sqlalchemy import (
    create_engine, MetaData, Table, Column,
    Integer, String, Float, ForeignKey
)

# --------------------------------------------------------------------
# 1) Load the cleaned tables (after Cleaning and Feature Engineering)
# --------------------------------------------------------------------
OUTPUT_DIR        = "/mnt/user-data/outputs"
dim_country       = pd.read_csv(f"{OUTPUT_DIR}/Dim_Country.csv")
fact_country_year = pd.read_csv(f"{OUTPUT_DIR}/Fact_Country_Year.csv")
fact_flows        = pd.read_csv(f"{OUTPUT_DIR}/Fact_Migration_Flows.csv")

# --------------------------------------------------------------------
# 2) Create a SQLite database (will be stored as a .db file on the
#    machine/Colab)
# --------------------------------------------------------------------
engine = create_engine("sqlite:///brain_drain_model.db")
metadata = MetaData()

# --------------------------------------------------------------------
# 3) Define Dim_Country with a primary key (PK)
# --------------------------------------------------------------------
dim_country_table = Table(
    "Dim_Country", metadata,
    Column("country_iso3", String, primary_key=True),
    Column("country_name", String),
    Column("region_name", String),
    Column("income_level_name", String),
)

# --------------------------------------------------------------------
# 4) Define Fact_Country_Year with a foreign key (FK) to Dim_Country
# --------------------------------------------------------------------
fact_country_year_table = Table(
    "Fact_Country_Year", metadata,
    Column("id", Integer, primary_key=True, autoincrement=True),
    Column("country_iso3", String, ForeignKey("Dim_Country.country_iso3")),
    Column("year", Integer),
    Column("population", Float),
    Column("netmigration", Float),
)

# --------------------------------------------------------------------
# 5) Define Fact_Migration_Flows with two foreign keys to the same
#    Dim_Country (from and to) - this is an example of a role-playing
#    dimension
# --------------------------------------------------------------------
fact_flows_table = Table(
    "Fact_Migration_Flows", metadata,
    Column("id", Integer, primary_key=True, autoincrement=True),
    Column("from_country_iso3", String, ForeignKey("Dim_Country.country_iso3")),
    Column("to_country_iso3", String, ForeignKey("Dim_Country.country_iso3")),
    Column("year", Integer),
    Column("n_migrations", Integer),
    Column("netmigrations", Integer),
)

# --------------------------------------------------------------------
# 6) Actually create the tables inside the database
# --------------------------------------------------------------------
metadata.create_all(engine)

# --------------------------------------------------------------------
# 7) Upload the data from the DataFrames into the database (the actual
#    linking)
#    Note: Dim_Country must be uploaded first, since the fact tables
#    depend on it
# --------------------------------------------------------------------
dim_country[["country_iso3", "country_name", "region_name", "income_level_name"]]. \
to_sql("Dim_Country", engine, if_exists="replace", index=False)

fact_country_year[["country_iso3", "year", "population", "netmigration"]]. \
to_sql("Fact_Country_Year", engine, if_exists="append", index=False)

fact_flows[["from_country_iso3", "to_country_iso3", "year", "n_migrations", "netmigrations"]]. \
to_sql("Fact_Migration_Flows", engine, if_exists="append", index=False)

print("✅ Database built and tables linked successfully: brain_drain_model.db")

# --------------------------------------------------------------------
# 8) Quick test to confirm the link works: a real JOIN between the fact
#    table and the dimension
# --------------------------------------------------------------------
test_query = """
SELECT f.year, f.population, d.country_name, d.region_name
FROM Fact_Country_Year f
JOIN Dim_Country d ON f.country_iso3 = d.country_iso3
LIMIT 5
"""
result = pd.read_sql(test_query, engine)
print("\nSample of the JOIN result between Fact_Country_Year and Dim_Country:")
print(result)


# **7- ERD Diagram**

In [ ]:
"""
==========================================================================
 Stage: Data Modeling - Code 2 (drawing the ERD Diagram)
==========================================================================
Goal: visually draw the Star Schema (as an image) from the same database
      we built in the previous code, so you can confirm the relationships
      are correct and use this image in the project's presentation or
      README.
==========================================================================
"""

from eralchemy2 import render_er

# Draws the ERD directly from the SQLite database file built in the previous code
render_er("sqlite:///brain_drain_model.db", "brain_drain_erd.png")

print("✅ ERD Diagram saved to: brain_drain_erd.png")


# **9- Data Analysis**

In [ ]:
OUTPUT_DIR = "/mnt/user-data/outputs"

"""
==========================================================================
 Code 1: Top 15 countries gaining researchers (Brain Gain)
==========================================================================
Goal: show the countries that gain more researchers than they lose
      (positive net migration), aggregated across all years (1996-2023)
      for each country.
==========================================================================
"""

import pandas as pd
import matplotlib.pyplot as plt

fact_country_year = pd.read_csv(f"{OUTPUT_DIR}/Fact_Country_Year.csv")
dim_country       = pd.read_csv(f"{OUTPUT_DIR}/Dim_Country.csv")

df = fact_country_year.merge(dim_country[["country_iso3", "country_name"]], on="country_iso3", how="left")

top_gain = (
    df.groupby("country_name")["netmigration"].sum()
    .sort_values(ascending=False).head(15).reset_index()
)

plt.figure(figsize=(10, 7))
plt.barh(top_gain["country_name"], top_gain["netmigration"], color="#2D6A4F")
plt.gca().invert_yaxis()  # show the top country at the top
plt.title("Top 15 Countries Gaining Researchers (Brain Gain)", fontsize=14)
plt.xlabel("Net Migration (total researchers, 1996–2023)")
plt.tight_layout()
plt.savefig("top_brain_gain_countries.png", dpi=150)
plt.show()

print("✅ Saved: top_brain_gain_countries.png")


In [ ]:
OUTPUT_DIR = "/mnt/user-data/outputs"

"""
==========================================================================
 Code 2: Top 15 countries losing researchers (Brain Drain)
==========================================================================
Goal: show the countries that lose more researchers than they gain
      (negative net migration), aggregated across all years (1996-2023)
      for each country.
==========================================================================
"""

import pandas as pd
import matplotlib.pyplot as plt

fact_country_year = pd.read_csv(f"{OUTPUT_DIR}/Fact_Country_Year.csv")
dim_country       = pd.read_csv(f"{OUTPUT_DIR}/Dim_Country.csv")

df = fact_country_year.merge(dim_country[["country_iso3", "country_name"]], on="country_iso3", how="left")

top_drain = (
    df.groupby("country_name")["netmigration"].sum()
    .sort_values().head(15).reset_index()
)

plt.figure(figsize=(10, 7))
plt.barh(top_drain["country_name"], top_drain["netmigration"], color="#BC4749")
plt.gca().invert_yaxis()  # show the biggest losing country at the top
plt.title("Top 15 Countries Losing Researchers (Brain Drain)", fontsize=14)
plt.xlabel("Net Migration (total researchers, 1996–2023)")
plt.tight_layout()
plt.savefig("top_brain_drain_countries.png", dpi=150)
plt.show()

print("✅ Saved: top_brain_drain_countries.png")


In [ ]:
OUTPUT_DIR = "/mnt/user-data/outputs"

"""
==========================================================================
 Code 3: How a country's economy is affected by researcher migration
 (Economic Impact)
==========================================================================
Note: the original data doesn't have an actual GDP column, so we used
      income_level_name (the World Bank classification: Low/Lower
      middle/Upper middle/High income) as a reliable alternative economic
      indicator for the same purpose.

Goal: show the average net migration for each income category, to
      demonstrate that higher-income countries gain researchers, while
      lower-income countries lose them.
==========================================================================
"""

import pandas as pd
import matplotlib.pyplot as plt

fact_country_year = pd.read_csv(f"{OUTPUT_DIR}/Fact_Country_Year.csv")
dim_country       = pd.read_csv(f"{OUTPUT_DIR}/Dim_Country.csv")

df = fact_country_year.merge(dim_country[["country_iso3", "income_level_name"]], on="country_iso3", how="left")
df = df.dropna(subset=["income_level_name"])
df = df[df["income_level_name"] != "Not classified"]

# Logical ordering of income levels instead of alphabetical order
income_order = ["Low income", "Lower middle income", "Upper middle income", "High income"]
avg_by_income = df.groupby("income_level_name")["netmigration"].mean().reindex(income_order)

colors = ["#BC4749", "#E9A178", "#74C69D", "#1B4332"]  # from red (loss) to dark green (gain)

plt.figure(figsize=(9, 6))
plt.bar(avg_by_income.index, avg_by_income.values, color=colors)
plt.axhline(0, color="black", linewidth=0.8)
plt.title("Average Net Migration by Country Income Level", fontsize=14)
plt.ylabel("Average Net Migration (researchers)")
plt.xlabel("Income Level")
plt.tight_layout()
plt.savefig("economic_impact_migration.png", dpi=150)
plt.show()

print("✅ Saved: economic_impact_migration.png")
print(avg_by_income)


In [ ]:
OUTPUT_DIR = "/mnt/user-data/outputs"

"""
==========================================================================
 Code 4: Line Chart - global researcher migration trend over the years
==========================================================================
Goal: show two lines on the same chart (total in-migrations and total
      out-migrations across all countries) for each year, to see whether
      global researcher mobility is increasing or decreasing over time.
==========================================================================
"""

import pandas as pd
import matplotlib.pyplot as plt

fact_country_year = pd.read_csv(f"{OUTPUT_DIR}/Fact_Country_Year.csv")

trend = fact_country_year.groupby("year")[["number_of_inmigrations", "number_of_outmigrations"]].sum().reset_index()

plt.figure(figsize=(11, 6))
plt.plot(trend["year"], trend["number_of_inmigrations"], color="#2D6A4F", linewidth=4, label="In-migrations", alpha=0.9)
plt.plot(trend["year"], trend["number_of_outmigrations"], color="#95D5B2", linewidth=2, linestyle="--", label="Out-migrations")
plt.title("Global Researcher Migration Trend Over Time", fontsize=14)
plt.xlabel("Year")
plt.ylabel("Total Migrations (all countries)")
plt.legend()
plt.tight_layout()
plt.savefig("global_migration_trend_line.png", dpi=150)
plt.show()

print("✅ Saved: global_migration_trend_line.png")
print("Note: the two lines are nearly identical because the global total of in-migrations equals the global total of out-migrations (every migration is counted once as in and once as out)")


In [ ]:
OUTPUT_DIR = "/mnt/user-data/outputs"

"""
==========================================================================
 Code 5: Pie Chart - share of records by migration status
 (Brain Gain/Drain/Balanced)
==========================================================================
Goal: show the share of each migration status (Brain Gain / Brain Drain /
      Balanced) out of all country-year records, to see the overall
      picture: are most cases gains or losses?
Note: the "Unknown" category was excluded because it represents missing
      data (years not yet updated), not an actual migration status.
==========================================================================
"""

import pandas as pd
import matplotlib.pyplot as plt

fact_country_year = pd.read_csv(f"{OUTPUT_DIR}/Fact_Country_Year.csv")

status_counts = fact_country_year["migration_status"].value_counts()
status_counts = status_counts.drop("Unknown", errors="ignore")

colors = {"Brain Gain": "#2D6A4F", "Brain Drain": "#BC4749", "Balanced": "#95D5B2"}
plot_colors = [colors[label] for label in status_counts.index]

plt.figure(figsize=(7, 7))
plt.pie(
    status_counts.values,
    labels=status_counts.index,
    autopct="%1.1f%%",
    colors=plot_colors,
    startangle=90,
    wedgeprops={"edgecolor": "white", "linewidth": 1.5}
)
plt.title("Share of Country-Year Records by Migration Status", fontsize=14)
plt.tight_layout()
plt.savefig("migration_status_pie.png", dpi=150)
plt.show()

print("✅ Saved: migration_status_pie.png")
print(status_counts)


In [ ]:
OUTPUT_DIR = "/mnt/user-data/outputs"

"""
==========================================================================
 Code 6: Map - world map of net researcher migration by country
==========================================================================
Goal: an interactive choropleth map coloring each country by its total
      net migration (aggregated across all years) - dark green = gaining
      researchers, red = losing them.
Note: we used country_iso3 directly as location codes since they match
      the ISO Alpha-3 standard that plotly understands automatically.
==========================================================================
"""

import pandas as pd
import plotly.express as px

fact_country_year = pd.read_csv(f"{OUTPUT_DIR}/Fact_Country_Year.csv")
dim_country       = pd.read_csv(f"{OUTPUT_DIR}/Dim_Country.csv")

# Aggregate total net migration for each country across all years
country_totals = fact_country_year.groupby("country_iso3")["netmigration"].sum().reset_index()
country_totals = country_totals.merge(dim_country[["country_iso3", "country_name"]], on="country_iso3", how="left")

fig = px.choropleth(
    country_totals,
    locations="country_iso3",
    color="netmigration",
    hover_name="country_name",
    color_continuous_scale="RdYlGn",   # red (loss) -> yellow (balanced) -> green (gain)
    color_continuous_midpoint=0,       # zero in the middle, so the gain/loss split is clear
    title="Net Researcher Migration by Country (1996–2023)"
)
fig.update_layout(margin={"r":0, "t":40, "l":0, "b":0})

fig.write_html("net_migration_world_map.html", include_plotlyjs="cdn")   # cdn instead of embedding the full library in the file (shrinks size from 4MB to under 50KB)

# --------------------------------------------------------------------
# Show a clickable link that opens the map in a new tab directly
# (without downloading the file first) - we convert the HTML content to
# base64 and put it as a data URI inside an <a> link
# --------------------------------------------------------------------
import base64
from IPython.display import HTML, display

with open("net_migration_world_map.html", "rb") as f:
    html_bytes = f.read()
encoded = base64.b64encode(html_bytes).decode()
data_uri = f"data:text/html;base64,{encoded}"

display(HTML(f'<a href="{data_uri}" target="_blank" style="font-size:16px; color:#2D6A4F; font-weight:bold;">🌍 Open the interactive map from here</a>'))

# Save a static image version (PNG) - requires Chrome to be installed in
# the runtime environment. If you get a Chrome error in Colab, run this
# line once in its own cell:
#   !plotly_get_chrome -y
try:
    fig.write_image("net_migration_world_map.png", width=1200, height=650, scale=2)
    print("✅ Saved: net_migration_world_map.png and net_migration_world_map.html")
except Exception as e:
    print("⚠️ Only net_migration_world_map.html was saved (the PNG version needs Chrome installed)")
    print("If you want a PNG version, run this command once in Colab: !plotly_get_chrome -y")


In [ ]:
OUTPUT_DIR = "/mnt/user-data/outputs"

"""
==========================================================================
 Code 7: Top 15 Migration Corridors - the largest real migration routes
 between countries
==========================================================================
Goal: show the top 15 "migration corridors" (from->to country pairs) by
      number of migrating researchers, aggregated across all years.

Important note: the value "UNKNOWN" was excluded from the source/
      destination, since it represents cases where "the other party is
      not known" rather than a real country - leaving it in would push it
      to the top of the results and distort the meaning of "the biggest
      migration corridor between two countries."
==========================================================================
"""

import pandas as pd
import matplotlib.pyplot as plt

fact_flows = pd.read_csv(f"{OUTPUT_DIR}/Fact_Migration_Flows.csv")
dim_country = pd.read_csv(f"{OUTPUT_DIR}/Dim_Country.csv")

# Exclude any corridor where UNKNOWN is the source or destination
real_flows = fact_flows[
    (fact_flows["from_country_iso3"] != "UNKNOWN") &
    (fact_flows["to_country_iso3"] != "UNKNOWN")
].copy()

# Aggregate each corridor across all years, and get country names instead of codes
name_map = dim_country.set_index("country_iso3")["country_name"].to_dict()
real_flows["from_name"] = real_flows["from_country_iso3"].map(name_map)
real_flows["to_name"] = real_flows["to_country_iso3"].map(name_map)
real_flows["corridor_label"] = real_flows["from_name"] + " → " + real_flows["to_name"]

top_corridors = (
    real_flows.groupby("corridor_label")["n_migrations"].sum()
    .sort_values(ascending=False).head(15).reset_index()
)

plt.figure(figsize=(10, 8))
plt.barh(top_corridors["corridor_label"], top_corridors["n_migrations"], color="#40916C")
plt.gca().invert_yaxis()
plt.title("Top 15 Migration Corridors Between Countries (1998–2020)", fontsize=14)
plt.xlabel("Total Number of Migrations")
plt.tight_layout()
plt.savefig("top_corridors.png", dpi=150)
plt.show()

print("✅ Saved: top_corridors.png")
print(top_corridors)


<h1 align="center">🌍 Global Scholarly Migration — Insights & Recommendations</h1>

<h2>🔍 Key Insights</h2>

<table>
<tr style="background-color:#1B4332; color:white;">
<th width="5%">#</th>
<th width="30%">Insight</th>
<th width="65%">Supporting Evidence</th>
</tr>

<tr style="background-color:#D8F3DC;">
<td align="center">1</td>
<td><b>🎯 Extreme Concentration, Not Broad Patterns</b></td>
<td>The <b>United States</b> alone gains +32,609 researchers (1996–2023) — 4x the runner-up (Switzerland, +13,757). <b>India</b> loses -25,745, ~1.5x the UK, the second-biggest loser.</td>
</tr>

<tr style="background-color:#F1F8F4;">
<td align="center">2</td>
<td><b>📊 Outliers Reflect Real Scale, Not Errors</b></td>
<td>Population & migration show ~19–26% statistical outliers (IQR). Confirmed via box plots/scatter — these are real countries (China, India, US), so raw values were kept unchanged.</td>
</tr>

<tr style="background-color:#D8F3DC;">
<td align="center">3</td>
<td><b>💰 High-Income Countries Attract Talent — With a Catch</b></td>
<td>Avg. net migration rises with income level overall. But the "Lower-middle income" group looks negative only because <b>India alone</b> skews it — not a general pattern for that tier.</td>
</tr>

<tr style="background-color:#F1F8F4;">
<td align="center">4</td>
<td><b>📉 Mobility Grew for 2 Decades, Then Collapsed</b></td>
<td>Global migration rose steadily from 1996, peaked ~2020, then dropped sharply — consistent with COVID-19 travel/border restrictions.</td>
</tr>

<tr style="background-color:#D8F3DC;">
<td align="center">5</td>
<td><b>🔀 A Handful of Corridors Dominate</b></td>
<td>Nearly all top 15 corridors involve the <b>United States</b>. Top corridor: US ↔ China (~54,000 each direction), followed by UK/Canada ↔ US.</td>
</tr>

<tr style="background-color:#F1F8F4;">
<td align="center">6</td>
<td><b>⚖️ More Country-Years Show Gains Than Losses</b></td>
<td>54.9% of country-year records show "Brain Gain" vs. 35.2% "Brain Drain" — but a few countries dominate the actual volume in each direction.</td>
</tr>

</table>

<br>

<h2>💡 Recommendations</h2>

<table>
<tr style="background-color:#40916C; color:white;">
<th width="5%">#</th>
<th width="30%">Recommendation</th>
<th width="65%">Why</th>
</tr>

<tr style="background-color:#B7E4C7;">
<td align="center">1</td>
<td><b>Focus on outlier countries, not averages</b></td>
<td>Migration is highly concentrated, so group averages (e.g., by income level) can mislead. Validate findings against individual countries like India, US, China.</td>
</tr>

<tr style="background-color:#E9F5EC;">
<td align="center">2</td>
<td><b>Investigate the post-2020 decline</b></td>
<td>The sharp drop after 2020 deserves dedicated analysis — ideally cross-referenced with a COVID-19 timeline — to separate a temporary shock from a lasting structural shift.</td>
</tr>

<tr style="background-color:#B7E4C7;">
<td align="center">3</td>
<td><b>Prioritize retention strategies for top "exporters"</b></td>
<td>India, UK, France, and Italy would benefit from case studies on *why* researchers leave, since India's loss alone drags down its entire income group's stats.</td>
</tr>

<tr style="background-color:#E9F5EC;">
<td align="center">4</td>
<td><b>Treat US-connected corridors as a separate sub-network</b></td>
<td>Since almost all top corridors involve the US, separating "US-linked flows" from "non-US flows" could reveal hidden regional hubs (e.g., within Europe or Asia).</td>
</tr>

<tr style="background-color:#B7E4C7;">
<td align="center">5</td>
<td><b>Improve data completeness for "Unknown" flows</b></td>
<td>A meaningful share of flow records have an unidentified origin/destination. Better source data would sharpen corridor-level totals and accuracy.</td>
</tr>

<tr style="background-color:#E9F5EC;">
<td align="center">6</td>
<td><b>Report medians alongside means</b></td>
<td>Since single-country outliers (like India) distort group averages, future dashboards should show median/trimmed-mean metrics for fairer cross-country comparisons.</td>
</tr>

</table>